src/magic_folder/test/test_schema.py

```diff
@@ -0,0 +1,192 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+"""
+Tests for ``_zkapauthorizer.schema``.
+"""
+
+from __future__ import (
+    unicode_literals,
+    absolute_import,
+)
+
+from testtools import (
+    TestCase,
+    ExpectedException,
+)
+from testtools.matchers import (
+    Equals,
+    MatchesStructure,
+)
+
+from hypothesis import (
+    given,
+)
+from hypothesis.strategies import (
+    integers,
+    data,
+    lists,
+)
+
+from sqlite3 import (
+    connect,
+)
+
+from .._schema import (
+    MAXIMUM_UPGRADES,
+    DatabaseSchemaTooNew,
+    SchemaUpgrade,
+    Schema,
+)
+
+class SchemaTests(TestCase):
+    """
+    Tests for ``Schema``.
+    """
+    @given(
+        integers(
+            min_value=MAXIMUM_UPGRADES + 1,
+            max_value=MAXIMUM_UPGRADES * 100,
+        ),
+    )
+    def test_too_many_upgrades(self, num_upgrades):
+        """
+        ``Schema`` raises `ValueError`` if initialized with a schema with more
+        than ``MAXIMUM_UPGRADES`` upgrades.
+        """
+        with ExpectedException(ValueError):
+            Schema(
+                upgrades=dummy_upgrades(num_upgrades),
+            )
+
+    @given(
+        integers(min_value=0, max_value=MAXIMUM_UPGRADES),
+    )
+    def test_version(self, num_upgrades):
+        """
+        ``Schema.version`` evaluates to the version that the schema itself
+        defines.
+        """
+        upgrades = dummy_upgrades(num_upgrades)
+        schema = Schema(upgrades=upgrades)
+        self.assertThat(
+            schema,
+            MatchesStructure(
+                version=Equals(num_upgrades),
+            ),
+        )
+
+    @given(
+        integers(min_value=0, max_value=MAXIMUM_UPGRADES),
+    )
+    def test_get_version(self, num_upgrades):
+        """
+        ``Schema.get_version`` returns the version number to which the schema has
+        been upgraded.
+        """
+        upgrades = dummy_upgrades(num_upgrades)
+        schema = Schema(upgrades=upgrades)
+
+        db = connect(":memory:")
+        cursor = db.cursor()
+        schema.run_upgrades(cursor)
+        self.assertThat(
+            schema.get_version(cursor),
+            Equals(num_upgrades),
+        )
+
+    @given(
+        integers(min_value=0, max_value=MAXIMUM_UPGRADES),
+        integers(min_value=1, max_value=2 ** 63),
+    )
+    def test_database_newer_than_schema(self, num_upgrades, additional_versions):
+        """
+        ``Schema.run_upgrades`` raises ``ValueError`` if initialized with a schema
+        with a version that is less than the version recorded in the database.
+        """
+        schema = Schema(upgrades=dummy_upgrades(num_upgrades))
+        db = connect(":memory:")
+        cursor = db.cursor()
+
+        # Force version schema creation.
+        schema.get_version(cursor)
```

Comment:
This seems surprising that a method called `get_version` actually creates table(s) in the database. Maybe a different method-name? Or separate "get the version" from "create a table with the version"?

Reply:
Yea, that was lame.  Switched to using the built-in "user_version" pragma which eliminates the need to create our own table for the version.